contracts/modules/dispute/DisputeModule.sol

```diff
@@ -239,19 +260,69 @@ contract DisputeModule is
         emit DisputeCancelled(disputeId, data);
     }
 
+    /// @notice Tags a derivative if a parent has been tagged with an infringement tag
+    /// @param parentIpId The infringing parent ipId
+    /// @param derivativeIpId The derivative ipId
+    /// @param parentDisputeId The dispute id that tagged the parent ipId as infringing
+    function tagDerivativeIfParentInfringed(
+        address parentIpId,
+        address derivativeIpId,
+        uint256 parentDisputeId
+    ) external {
+        DisputeModuleStorage storage $ = _getDisputeModuleStorage();
+
+        Dispute memory parentDispute = $.disputes[parentDisputeId];
+        if (parentDispute.targetIpId != parentIpId) revert Errors.DisputeModule__ParentIpIdMismatch();
+        if (parentDispute.currentTag == IN_DISPUTE || parentDispute.currentTag == bytes32(0))
+            revert Errors.DisputeModule__ParentNotTagged();
+
+        if (!LICENSE_REGISTRY.isParentIp(parentIpId, derivativeIpId)) revert Errors.DisputeModule__NotDerivative();
+
+        address arbitrationPolicy = $.arbitrationPolicies[derivativeIpId];
+        if (!$.isWhitelistedArbitrationPolicy[arbitrationPolicy]) arbitrationPolicy = $.baseArbitrationPolicy;
+
+        uint256 disputeId = ++$.disputeCounter;
+
+        $.disputes[disputeId] = Dispute({
+            targetIpId: derivativeIpId,
+            disputeInitiator: msg.sender,
+            arbitrationPolicy: arbitrationPolicy,
+            linkToDisputeEvidence: "",
+            targetTag: parentDispute.currentTag,
+            currentTag: parentDispute.currentTag,
+            parentDisputeId: parentDisputeId
+        });
+
+        $.successfulDisputesPerIp[derivativeIpId]++;
+
+        emit DerivativeTaggedOnParentInfringement(
+            parentIpId,
+            derivativeIpId,
+            parentDisputeId,
+            parentDispute.currentTag
+        );
+    }
+
     /// @notice Resolves a dispute after it has been judged
     /// @param disputeId The dispute id
-    function resolveDispute(uint256 disputeId) external {
+    /// @param data The data to resolve the dispute
+    function resolveDispute(uint256 disputeId, bytes calldata data) external {
         DisputeModuleStorage storage $ = _getDisputeModuleStorage();
         Dispute memory dispute = $.disputes[disputeId];
 
-        if (msg.sender != dispute.disputeInitiator) revert Errors.DisputeModule__NotDisputeInitiator();
+        if (dispute.parentDisputeId == 0 && msg.sender != dispute.disputeInitiator)
```

Comment:
Why check `dispute.parentDisputeId == 0`? 

If it's 1 or more, the branch skips on checking if msg.sender is the dispute initiator.

Reply:
Yes, so there are 2 types of dispute now:

1. the normal one that requires judgment: this one only the dispute initiator can resolve
2. the automatic one: this one is permissionless to tag and permissionless to resolve

the way to distinguish between them is whether the `dispute.parentDisputeId == 0`

Reply:
you need to add better comments here to avoid confusion later, especially this is a key business logic. thanks.

Reply:
addressed in https://github.com/storyprotocol/protocol-core-v1/pull/60/commits/6581549f82c44095d31241a4ac7c12f00dc13ddb